README.md
# Arduino USB Host Shield 2.0 MIDI Transport
[![arduino-library-badge](https://www.ardu-badge.com/badge/UHS2-MIDI.svg?)](https://www.ardu-badge.com/UHS2-MIDI)
[![Check Arduino status](https://github.com/YuuichiAkagawa/Arduino-UHS2MIDI/actions/workflows/check-arduino.yml/badge.svg)](https://github.com/YuuichiAkagawa/Arduino-UHS2MIDI/actions/workflows/check-arduino.yml")
[![License](https://img.shields.io/github/license/YuuichiAkagawa/Arduino-UHS2MIDI.svg?maxAge=3600)](LICENSE)

USB Host Shield 2.0  MIDI transport layer for the [FortySevenEffects Arduino MIDI Library](https://github.com/FortySevenEffects/arduino_midi_library) and uses the underlying [USB Host Shield 2.0](https://github.com/felis/USB_Host_Shield_2.0) library.  
This library is based on the [Arduino-USBMIDI](https://github.com/lathoub/Arduino-USBMIDI).

## Installation
Install from Arduino Library Manager. Search for UHS2-MIDI.
![arduino-library-manager](https://user-images.githubusercontent.com/926923/117981071-92c41d80-b36f-11eb-9cd1-30423f708880.png)

## Usage
### Basic / Default
```cpp
#include <UHS2-MIDI.h>
...
USB Usb;
UHS2MIDI_CREATE_DEFAULT_INSTANCE(&Usb);
...
void setup()
{
  MIDI.begin(1);
  if (Usb.Init() == -1) {
    while (1); //halt
  }//if (Usb.Init() == -1...
  ...
}
void loop()
{
  Usb.Task();
  MIDI.read();
  ...
```
will create a instance named `MIDI` (transport instance named `__uhs2MIDI`) and is by default connected to cable number 0 - and listens to incoming MIDI on channel 1.

### Modified
```cpp
#include <UHS2-MIDI.h>
...
USB Usb;
UHS2MIDI_CREATE_INSTANCE(&Usb, 4, MIDI);
```
will create a instance named `MIDI` (transport instance named `__uhs2MIDI`) and is connected to cable number 4.

### Custom Settings
```cpp
#include <UHS2-MIDI.h>
...
struct MySettings : public midi::DefaultSettings
{
    static const unsigned SysExMaxSize = 512; // Accept SysEx messages up to 512 bytes long.
};
USB Usb;
UHS2MIDI_CREATE_CUSTOM_INSTANCE(&Usb, 0, MIDI, MySettings);
```
will create a instance named `MIDI` (transport instance named `__uhs2MIDI`) and change the maximum size of SysEx messages to 512 bytes.
### Advanced
```cpp
#include <UHS2-MIDI.h>
...
USB Usb;
UHS2MIDI_NAMESPACE::uhs2MidiTransport uhs2MIDI2(&Usb, 5);
MIDI_NAMESPACE::MidiInterface<UHS2MIDI_NAMESPACE::uhs2MidiTransport> MIDI2((UHS2MIDI_NAMESPACE::uhs2MidiTransport&)uhs2MIDI2);
```
will create a instance named `uhs2MIDI2` (and underlaying MIDI object `MIDI2`) and is by default connected to cable number 5.

### Change CableNumber
```cpp
#include <UHS2-MIDI.h>
...
USB Usb;
UHS2MIDI_CREATE_DEFAULT_INSTANCE(&Usb);
void setup()
{
  MIDI.begin(4);                            // Launch MIDI and listen to channel 4
  MIDI.getTransport()->setCableNumber(1);   // Change CableNumber to 1 later
```
## Tested boards / modules
- Arduino UNO R3

## Changelog
[Chagelog](/Changelog.md)

## Other Transport protocols:
The libraries below  the same calling mechanism (API), making it easy to interchange the transport layer.
- [Arduino USB-MIDI  Transport](https://github.com/lathoub/USB-MIDI)
- [Arduino AppleMIDI Transport](https://github.com/lathoub/Arduino-AppleMIDI-Library)
- [Arduino ipMIDI  Transport](https://github.com/lathoub/Arduino-ipMIDI)
- [Arduino BLE-MIDI  Transport](https://github.com/lathoub/Arduino-BLE-MIDI)
- [Arduino USB Host Library SAMD MIDI Transport](https://github.com/YuuichiAkagawa/Arduino-USBHSAMD-MIDI)


Changelog.md
# Changelog
## 2023.08.12(1.0.1)
- Add new API ```void setCableNumber(uint8_t cn)```

## 2021.05.10(1.0.0)
- GA
- update document

## 2021.05.06(0.0.3)
- Add UHS2MIDI_CREATE_CUSTOM_INSTANCE macro

## 2021.05.02(0.0.2)
- Refactaring: improve performance and reduce SRAM usage

## 2021.04.30(0.0.1)
- initial release


src/UHS2-MIDI_defs.h
/*!
 *  @file       UHS2-MIDI_defs.h
 *  Project     Arduino UHS2 MIDI transport
 *  @brief      UHS2 MIDI transport for the Arduino - Definitions
 *  @author     Yuuichi Akagawa
 *  @date       20/12/2020
 *  @license    MIT - Copyright (c) 2020-2021 Yuuichi Akagawa
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */

#pragma once

#include "UHS2-MIDI_Namespace.h"

BEGIN_UHS2MIDI_NAMESPACE

//USB-MIDI event packet
typedef union
{
  uint8_t buf[4];
  struct{
    uint8_t header;
    uint8_t byte1;
    uint8_t byte2;
    uint8_t byte3;
  } p;
} usbMidiEventPacket_t;

using namespace MIDI_NAMESPACE;

const uint8_t Fsys2cin[] PROGMEM =  {0, 2, 3, 2, 0, 0, 5, 0, 0xf, 0, 0xf, 0xf, 0xf, 0, 0xf, 0xf};

#define MAKEHEADER(cn, cin) (((cn & 0x0f) << 4) | cin)

END_UHS2MIDI_NAMESPACE


src/UHS2-MIDI_Namespace.h
/*!
 *  @file       UHS2-MIDI_Namespace.h
 *  Project     Arduino UHS2 MIDI transport
 *  @brief      UHS2 MIDI transport for the Arduino - Namespace declaration
 *  @author     Yuuichi Akagawa
 *  @date       20/12/2020
 *  @license    MIT - Copyright (c) 2020 Yuuichi Akagawa
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */

#pragma once

#define UHS2MIDI_NAMESPACE         uhs2Midi
#define BEGIN_UHS2MIDI_NAMESPACE   namespace UHS2MIDI_NAMESPACE {
#define END_UHS2MIDI_NAMESPACE     }

#define USING_NAMESPACE_UHS2MIDI   using namespace UHS2MIDI_NAMESPACE;

BEGIN_UHS2MIDI_NAMESPACE

END_UHS2MIDI_NAMESPACE


src/UHS2-MIDI.h
/*
 Copyright (c) 2020 lathoub

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
 */
/*
 *******************************************************************************
 * Original version: https://github.com/lathoub/Arduino-USBMIDI/blob/master/src/USB-MIDI.h
 *
 * Modified for the USB Host Shield 2.0 + USBH_MIDI
 * copyright (c) 2020-2023 Yuuichi Akagawa
 *******************************************************************************
 */
#pragma once

#include <MIDI.h>
#include <usbh_midi.h>

#include "UHS2-MIDI_defs.h"
#include "UHS2-MIDI_Namespace.h"

#define UHS2_MIDI_VERSION 10001

BEGIN_UHS2MIDI_NAMESPACE
class uhs2MidiTransport : public USBH_MIDI
{
private:
    uint8_t mTxIndex;
    MidiType mTxStatus;
    usbMidiEventPacket_t mTxPacket;
    uint8_t mRxLength;
    uint8_t mRxIndex;
    usbMidiEventPacket_t mRxPacket;

    uint8_t cableNumber;

public:
    uhs2MidiTransport(USB *p, uint8_t cableNumber = 0) : USBH_MIDI(p)
    {
      this->cableNumber = cableNumber;
    };

    static const bool thruActivated = false;

    uint8_t SendData(usbMidiEventPacket_t &packet){
        if( !bPollEnable ) return 4;
        return USBH_MIDI::SendRawData(4, packet.buf);
    };

    uint8_t RecvData(usbMidiEventPacket_t &packet){
      uint8_t rc = USBH_MIDI::RecvData(packet.buf, true);
      if( rc == 0 ) {
        packet.p.header = 0;
      }
      return rc;
    };

    void begin()
    {
      mTxIndex = 0;
      mRxIndex = 0;
      mRxLength = 0;
    };

    bool beginTransmission(MidiType status)
    {
      mTxStatus = status;

      byte cin = 0;
      if (status < SystemExclusive) {
        // Non System messages
        cin = ((status & 0xF0) >> 4);
        mTxPacket.p.header = MAKEHEADER(cableNumber, cin);
      }
      else {
        // Only System messages
        cin = pgm_read_byte_near(Fsys2cin + (status & 0x0F));
        mTxPacket.p.header = MAKEHEADER(cableNumber, cin);
      }

      mTxPacket.p.byte1 = mTxPacket.p.byte2 = mTxPacket.p.byte3  = 0;
      mTxIndex = 0;

      return true;
    };

    void write(byte byte)
    {
      if (mTxStatus != MidiType::SystemExclusive) {
        mTxPacket.buf[mTxIndex+1] = byte;
      }
      else if (byte == MidiType::SystemExclusiveStart) {
        mTxPacket.p.header = MAKEHEADER(cableNumber, 0x04);
        mTxPacket.p.byte1 = byte;
      }
      else // SystemExclusiveEnd or SysEx data
      {
        if (byte == MidiType::SystemExclusiveEnd)
          mTxPacket.p.header = MAKEHEADER(cableNumber, (0x05 + mTxIndex));

        switch(mTxIndex){
          case 0:
            mTxPacket.p.byte1 = byte;
            mTxPacket.p.byte2 = mTxPacket.p.byte3 = 0x00;
            break;
          case 1:
            mTxPacket.p.byte2 = byte;
            mTxPacket.p.byte3 = 0x00;
            break;
          case 2:
            mTxPacket.p.byte3 = byte;
            if (byte != MidiType::SystemExclusiveEnd)
              SendData(mTxPacket);
            break;
          default:
            break;
        }
      }
      mTxIndex = (mTxIndex == 2) ? 0 : mTxIndex+1;
    };

    void endTransmission()
    {
      SendData(mTxPacket);
    };

    byte read()
    {
      auto byte = mRxPacket.buf[mRxIndex++];
      mRxLength--;
      return byte;
    };

    unsigned available()
    {
      // consume mRxBuffer first, before getting a new packet
      if (mRxLength > 0)
        return mRxLength;

      mRxIndex = 1;
      auto len = (size_t)RecvData(mRxPacket);
      if (mRxPacket.p.header != 0) {
        //auto cn  = GETCABLENUMBER(mRxPacket);
        auto cn = (mRxPacket.p.header >> 4) & 0x0f;
        if (cn != cableNumber)
          return 0;
        mRxLength = len;
      }
      return mRxLength;
    };

    void setCableNumber(uint8_t cn){
        if( cn > 15 )
            return;
        cableNumber = cn;
    };
};

END_UHS2MIDI_NAMESPACE

#define UHS2MIDI_CREATE_INSTANCE(USBH, CableNr, Name)  \
  UHS2MIDI_NAMESPACE::uhs2MidiTransport __uhs2##Name(USBH, CableNr);\
  MIDI_NAMESPACE::MidiInterface<UHS2MIDI_NAMESPACE::uhs2MidiTransport> Name((UHS2MIDI_NAMESPACE::uhs2MidiTransport&)__uhs2##Name);

#define UHS2MIDI_CREATE_DEFAULT_INSTANCE(USBH)  \
  UHS2MIDI_CREATE_INSTANCE(USBH, 0, MIDI)

#define UHS2MIDI_CREATE_CUSTOM_INSTANCE(USBH, CableNr, Name, Settings)  \
  UHS2MIDI_NAMESPACE::uhs2MidiTransport __uhs2##Name(USBH, CableNr);\
  MIDI_NAMESPACE::MidiInterface<UHS2MIDI_NAMESPACE::uhs2MidiTransport, Settings > Name((UHS2MIDI_NAMESPACE::uhs2MidiTransport&)__uhs2##Name);
